### PR TITLE
Collections: remove redundant calls to .seq

### PR DESCRIPTION
**Pull request by retronym**
Students of Scala's history might recall that at introduction of
parallel collections, GenIterable et al were _not_ added; instead
the parallel collections inherited from the existing interfaces.

This of course was an invitation to widespread disaster as any
existing code that foreach-ed over a collection might now experience
unwanted concurrency.

The first attempt to fix this was to add the `.seq` method to
convert a parallel colleciton to a sequential one. This was added
in e579152f732, and call sites in the standard library with
side-effecting foreach calls were changed to use that.

But this was (rightly) deemed inadequate, as we could hardly expect
people to change existing code or remember to do this in new code.

So later, in 3de96153e5b, GenIterable was sprouted, and parallel
collections were re-parented.

This commit removes residual needless calls to .seq when the static
type of the receiver is already a plain Iterable, which are no-ops.

Review by @axel22

### CODE REVIEW

File: src/library/scala/collection/TraversableOnce.scala [hunk 1]
```diff
@@ -97,7 +97,7 @@ trait TraversableOnce[+A] extends Any with GenTraversableOnce[A] {
   // for internal use
   protected[this] def reversed = {
     var elems: List[A] = Nil
-    self.seq foreach (elems ::= _)
+    self foreach (elems ::= _)
```
Comment (axel22) [inline src/library/scala/collection/TraversableOnce.scala:100]: Seems like `this` should work as well.

**Comment by axel22** [comment]
LGTM. Strange that this wasn't noticed earlier.

File: src/library/scala/collection/TraversableOnce.scala (hunk 1 above)
Comment (Ichoran) [inline src/library/scala/collection/TraversableOnce.scala:100]: Whoa, whoa, that won't work without seq, will it?  elems isn't
synchronized.  If you traverse in parallel you'll end up with missed
elements.

Where are these pull requests?

  --Rex

Comment (retronym) [inline src/library/scala/collection/TraversableOnce.scala:100]: We know that `self` is already `TraversableOnce`, not a `GenTraversableOnce`.

Comment (Ichoran) [inline src/library/scala/collection/TraversableOnce.scala:100]: Ah, okay.  Fair enough then; I agree there's no need for .seq.

  --Rex

Comment (retronym) [inline src/library/scala/collection/TraversableOnce.scala:100]: @Ichoran 

The other pull requests will be trickling in. I've been throwing a lot of ideas around in https://github.com/retronym/scala/compare/topic/opt2 and will be working to find what's worthwhile in there with @gkossakowski next week.

On that branch, I saw improvements in the order of 10-15% of a warmed up resident compiler FSC repeatedly processing `src/library/util/**.scala`. But the usual caveats apply; that result might not generalize, and I might have made things incorrect along the way.

Some (many) of the commits there seemed to deliver cleaner YourKit profiles without real improvements in elapsed time.

Some profile results suprised me (as usual).  A lot of time was reported in the constructors of `AbstractSeq` and friends. I ended up writing a custom `ListBuffer` that didn't participate in the collections hierarchy at all to, and use that in a custom version of `map` that was specialized for `List` (no allocations on Nil.map, no indirection through the CanBuildFrom).

As soon as we figure out which changes seem to be the most beneficial, I'll let you know.

**Comment by Ichoran** [comment]
LGTM also.

**Comment by Ichoran** [comment]
I really need to get Jeremy Manson's async stack trace capturer (I hesitate to call it a profiler; it's very low level) working at a level where it could be useful for profiling the compiler.  I'm not sure whether YourKit uses AsyncGetCallTrace; it didn't used to because it wasn't part of the JVMTI specification.  But you can have huge discrepancies between what you profile and what actually is taking the time unless you use it.

**Comment by retronym** [comment]
@Ichoran I did try once to get the Oracle profiler up (http://www.oracle.com/technetwork/server-storage/solarisstudio/overview/index.html) which I've heard peeks under the hood to avoid the safepoint sampling bias and also gives some insights into JIT. But I got stuck and set it aside.

**Pull request merged**